### ISSUE TITLE
proposal: enforce check constraints on geometry validity

### ISSUE THREAD

**Issue by tudorbarascu** [qwat/qwat-data-model#102]
Today I found some errors on the subscribers geometries (not valid) and I remembered that I also had in the past invalid geometries on valves, pipes etc. 

Although this happens very rarely, I wanted to ask you @3nids and others if you ever considered enforcing geometry validity on the geometry columns and if you have any thoughts on this topic.

**Comment by tudorbarascu** [comment]
It would seem that some of the geometries that I had problems at were NULL, so a NOT NULL check would maybe mitigate this issue.

**Comment by 3nids** [comment]
I think that NULL is not an invalid geometry. Invalid geometry would rather be like a polygon with a ring that crosses itself.
I have no idea about forcing validity on the column.
Is there a common usage on this?

@vmora thoughts?

**Comment by tudorbarascu** [comment]
yes, NULL is not an invalid geometry.
I just realized that probably some of my coworkers probably imported some pipes, valves without geometry (through DB manager). I'll make a pull request to make sure the pipes valves etc have geometry.
I never thought this could happen but looking at the data in production it's clearly happening.

**Comment by 3nids** [comment]
well, then it makes sense to have not null geom for nodes and pipes I suppose.

**Comment by tudorbarascu** [comment]
Also a possible issue that may break our future functions on pipes (calculate height of vertexes with  interpolation) etc.:
Sometimes, during the pipe editing it happens that two vertexes are created at the same location.
I can't envision right now what it would break but I think that a check constraint to avoid this type of problem would be only beneficial.

**Comment by tudorbarascu** [comment]
We could also use ST_RemoveRepeatedPoints on the pipe geometry

**Comment by 3nids** [comment]
yes that could be added to the trigger

**Comment by tudorbarascu** [comment]
Do you think that we should ensure geometries don't have duplicates?

I found some in my data and I'm keen on not having them ever again.

I tried in https://github.com/qwat/qwat-data-model/pull/115 but the UNIQUE constraint is using BBOX to check the geometries,  which means that if two different geometries travelling somewhat the same path and having start end end points the same can have the same BBOX and they will colide with the UNIQUE constraint:
e.g.
```
test=# INSERT INTO qwat_od.pipe (id, fk_function, fk_installmethod, fk_material, fk_distributor, fk_precision, fk_bedding, fk_protection, fk_status, fk_watertype,  geometry)
test-# VALUES (2, 4105, 4201, 5121, 1, 1103, 9701, 4301, 1301, 1502, ST_GeomFromEWKT('SRID=21781;LINESTRING(500000 500000 0,500020 500000 0,500010 500010 0)'));
INSERT 0 1
test=# INSERT INTO qwat_od.distributor (name) VALUES ('Test Distributor');
INSERT 0 1
test=# 
test=# INSERT INTO qwat_od.pipe (id, fk_function, fk_installmethod, fk_material, fk_distributor, fk_precision, fk_bedding, fk_protection, fk_status, fk_watertype,  geometry)
test-# VALUES (1, 4105, 4201, 5121, 1, 1103, 9701, 4301, 1301, 1502, ST_GeomFromEWKT('SRID=21781;LINESTRING(500000 500000 0,500010 500000 0,500010 500010 0)'));
INSERT 0 1
test=# 
test=# INSERT INTO qwat_od.pipe (id, fk_function, fk_installmethod, fk_material, fk_distributor, fk_precision, fk_bedding, fk_protection, fk_status, fk_watertype,  geometry)
test-# VALUES (2, 4105, 4201, 5121, 1, 1103, 9701, 4301, 1301, 1502, ST_GeomFromEWKT('SRID=21781;LINESTRING(500000 500000 0,500005 500000 0,500010 500010 0)'));
ERROR:  duplicate key value violates unique constraint "pipe_geometry_key"
DETAIL:  Key (geometry)=(01020000A015550000030000000000000080841E410000000080841E4100000000000000000000000094841E410000000080841E41000000000000000000000000A8841E4100000000A8841E410000000000000000) already exists.
```
Also, the BBOX checking ignores the altitude of the constituent points.
I can think of different ways of doing it, but I'm not sure for now which one has better performance.
Any ideas that could save time?

**Comment by tudorbarascu** [comment]
While revisiting this I noticed again that the editors are doing copy and moving but sometimes they forget to move the pipe or they do it twice so we end up with duplicate geometries. So, unique geometry should be enforced.

For very long pipes, while trying to add the geom unique constraint I stumbled upon:
``
ERROR:  index row size 7400 exceeds maximum 2712 for index "geom_unique"
HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.
``
Of course, we do have some really long lines/pipes (more than 4 km..), if I split them etc. the constraint works. It would be nice not having to do this so any advice is welcome.

**Comment by haubourg** [comment]
Hi @tudorbarascu ! 
interesting. Can you post here the constraint DDL definition? It seems you use a btree index named "geom_unique", that is not in QWAT core. It seems btree indexes can't handle data with more that 2713 bytes. So I suppose  you do a btree index on the real geometry, which is especially expensive I think. 
I think a constraint checking first the spatial index identity based on the bbox comparison, then computing a real geometry equality test only if bboxes match would be faster and would avoid building a huge index.  

Or another option could be to build an index of a computed hash. It seems this is what is done here : https://smathermather.com/2009/10/22/unique-constraint-on-geometry-continued/

**Comment by tudorbarascu** [comment]
Hi @haubourg  For now it seems ``CREATE UNIQUE INDEX geom_unique ON qwat_od.pipe (md5(geometry::text));`` works. The last two hours I was testing if there are any performance implications and they are none.
E.g. doing a INSERT.. by copying all the pipes and translating them with 1 meter over x, y so that they won't duplicate: ``INSERT INTO qwat_od.pipe (fk_function, fk_installmethod, fk_material, fk_distributor, fk_precision, fk_bedding, fk_status, fk_watertype, geometry) SELECT fk_function, fk_installmethod, fk_material, fk_distributor, fk_precision, fk_bedding, fk_status, fk_watertype, ST_Translate(geometry, 1,1) FROM qwat_od.pipe;``

Has the same result with the unique index and without the index. I tested multiple times to be sure.
E.g.
* 915237.952 ms (15:15.238) no index
* 917398.427 ms (15:17.398) with index

**Comment by tudorbarascu** [comment]
Also, there are at least one duplicate in the demo data model as I restored it just now and...:
```
qwat=# ALTER TABLE qwat_od.pipe ADD CONSTRAINT geom_unique UNIQUE (geometry);
ERROR:  could not create unique index "geom_unique"
DETAIL:  Key (geometry)=(01020000A01555000002000000CFD322CC48731F410448D3BC81550341000000000000000028409D0547731F41247EB27A7E5503410000000000000000) is duplicated.
qwat=# SELECT id, ST_asText(geometry) FROM qwat_od.pipe WHERE geometry = '01020000A01555000002000000CFD322CC48731F410448D3BC81550341000000000000000028409D0547731F41247EB27A7E5503410000000000000000';
  id   |                                       st_astext                                       
-------+---------------------------------------------------------------------------------------
 32074 | LINESTRING Z (515282.199351606 158384.217199862 0,515281.755482676 158383.80991076 0)
 32077 | LINESTRING Z (515282.199351606 158384.217199862 0,515281.755482676 158383.80991076 0)
(2 rows)
```

The more reason why we should enforce that geometries are unique.

**Comment by haubourg** [comment]
@tudorbarascu agreed, duplicate nodes shouldn't be allowed. It is also related to https://github.com/qwat/QWAT/issues/169 that we postponed because I think we need handle node merging in a clear API more than in nested specialized triggers.
